integrations: Create incoming webhook for Airbyte.

aniebietafia opened the pull request:
This PR Resolves #31746 issue. It introduces a new integration tool - "Airbyte" in the Zulip project. The changes include adding the integration to the list of available integrations, creating documentation, adding fixtures for testing, and implementing the webhook view and tests.

PieterCK commented:
@aniebietafia It’s better to include fixes in the same commit where the issue happened, rather than adding new commits just to fix things. You can rebase them—this doc has some super helpful tips on how to fix git commits: https://zulip.readthedocs.io/en/latest/git/fixing-commits.html 👍

aniebietafia commented:
@PieterCK thank you for your earlier comment. I have rebased and made the commits more cleaner than before. I really love this. Please, let me know your thoughts and if there are more refinements to be made.

timabbott commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,87 @@
+    "Records Emitted: {recordsEmitted}, Records Committed: {recordsCommitted}. "
+    "Started at: {startedAt}, Finished at: {finishedAt}. "
+    "Bytes Emitted: {bytesEmitted}, Bytes Committed: {bytesCommitted}. "
+    "Duration In Seconds: {durationInSeconds}"
```
Can you use Zulip's formatting to make this easier to read and more useful?

I don't know much about how Airbyte is used, but I'm pretty sure that we can remove useless/duplicate information. Finish time is almost always going to be now, so I'd be tempted to do something like:

`Job {jobId} succeeded in {DurationFormatted}." with maybe some details in a bulleted list.

https://docs.airbyte.com/cloud/managing-airbyte-cloud/manage-airbyte-cloud-notifications#enable-the-slack-notification-in-airbyte

Has an example of what the Airbyte folks thought useful to put into a chat notification, and may be a useful guide.

aniebietafia replied:
Thank you for your feedback. I'll work on the changes. Please, can you point me to Zulip's formatting sample? This is my first contribution to this codebase. It will be faster and easier for me if I know where to look. Thank you.

laurynmm replied:
You can look at how some other existing webhook notification templates are formatted. But mostly @timabbott was referring to Zulip's Markdown formatting for messages. So newlines, bulleted lists, bold/italic text, etc.

If you look at the screenshot you generated, it's hard to read the message as it's formatted. In the dev environment, you can go to /devtools/integrations and generate some messages from different existing integrations to see how the Zulip messages appear.

aniebietafia replied:
Successfully refactored the code to make it more readable, implementing the following as pointed out:
* Formatted the responses using Zulip's Markdown formatting to make the responses more readable.
* Stripped out duplicate information from the response.

laurynmm commented:
You're going to want to fix your commit history, in particular the [merge commit in the series](https://zulip.readthedocs.io/en/latest/git/troubleshooting.html). For implementing a new webhook, you can do it in just one commit, which should be easier for you to maintain and will be fine for us to review.

You'll also want to use the tool to generate your screenshot for the integration, see bullet point about `generate-integration-docs-screenshot`: https://zulip.readthedocs.io/en/latest/documentation/integrations.html.

Finally, for the message template constant, I'd use the `"""` notation as it's easier to read. The changes to the templates for the GoCD view code in #31469 are a good example of that.

pranasziaukas commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,80 @@
+    "* **Records Emitted:** {recordsEmitted}\n"
+    "* **Records Committed:** {recordsCommitted}"
+)
+AIRBYTE_ERROR_MESSAGE_TEMPLATE = "**Job {jobId} failed.**\n" "* **Error Message:** {errorMessage}"
```
I don't think it's right to skip all of the details and only show the error message here, especially since there are no links to follow.

aniebietafia replied:
I get your message @pranasziaukas. So, in the case of an error during sync, the details of the job should still be displayed but with an error message attached. Right?

pranasziaukas replied:
Exactly, @aniebietafia 💯

aniebietafia replied:
@pranasziaukas I have incorporated the changes. Do mind reviewing this PR whenever you have the time? Thanks

pranasziaukas reviewed (changes requested):
As an Airbyte user, I would want to see most of the readable non-duplicate information from the payload.

Also, please note that the payloads signalling success and failure are almost identical -- the only difference is the presence/absense of the `errorMessage` field.
Therefore, I think it makes a lot of sense to use a single template and just include the error message if it's there.

Personally, I would use the following message template:
```
 Airbyte job {jobId} **{status}** in {durationFormatted}.

 Connection: [{connection_name}]({connection_url})
 Details:
 - Source: [{source_name}]({source_url})
 - Destination: [{destination_name}]({destination_url})
 - Records: {recordsEmitted} emitted, {recordsCommitted} committed
 - Bytes: {bytesEmittedFormatted} emitted, {bytesCommittedFormatted} committed
```
and append
```
- Error message: {errorMessage}
```
as needed.

If we want to follow [the Slack example](https://docs.airbyte.com/cloud/managing-airbyte-cloud/manage-airbyte-cloud-notifications#enable-the-slack-notification-in-airbyte), the template might be more like:
```
Airbyte sync {status}: [{connection_name}]({connection_url})

**Source:** [{source_name}]({source_url})
**Destination:** [{destination_name}]({destination_url})
**Duration:** {durationFormatted}

**Sync summary:**
{recordsEmitted} records emitted / {recordsCommitted} records committed
{bytesEmittedFormatted} emitted / {bytesCommittedFormatted} committed
```
(also with error message appended if applicable).

Both would essentially show the same information, so it's a matter of taste and preference.

aniebietafia commented:
Successfully implemented the following changes:
- Fixed commit history and unified to just one commit.
- Used this tool `./tools/screenshots/generate-integration-docs-screenshot --integration airbyte` to generate the response screenshot.
- Refactored the constants to use `multiline string formatting` making the code more readable.
- Refactored the code to append the `errorMessage` to the message template only when there is an error during sync.
- Formatted the Connection, Source and Destination to include the url links.

laurynmm commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,90 @@
+Airbyte job {jobId} **{status}** in {durationFormatted}.
+
+Connection: [{connection_name}]({connection_url})
+**Details:**
```
I'd drop this header and the job ID. Maybe simplify a bit and add an emoji green/red circle...

Example Message:

> :green_circle: Airbyte sync **succeeded** for [connection name link]().
> 
> - **Source**: [link]()
> - **Destination**: [link]()
>  - **Duration**:  time
> - **Records**: data
> - **Bytes**: data

aniebietafia replied:
Green if it is successful and red if the sync failed?

aniebietafia replied:
So, should take the duration out from the successful message and just add it as part of the details?

laurynmm replied:
Yeah, you can use the emoji strings for the two circles: `":green_circle:"` and `""red_circle"`.

And I think the duration would be better as a bullet point in the list. I think the more important part of the success/fail message is the link to the connection that's set up in Airbyte and the status.

laurynmm reviewed (commented):
@aniebietafia - Thanks for working on this integration! I left some feedback on the documentation and the view code. Let me know if you have any questions about my  comments.

Also, for the fixture data, these look like the [examples in Airbyte's documentation](https://docs.airbyte.com/cloud/managing-airbyte-cloud/manage-airbyte-cloud-notifications#example-webhook-notification-payload) on webhook notifications. I think our general preference is to have the example fixture data generated directly from the application we're integrating with and not taken from the application's documentation. @timabbott - Can you confirm whether that's correct or not?

aniebietafia commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,90 @@
+from zerver.lib.webhooks.common import check_send_webhook_message
+from zerver.models import UserProfile
+
+AIRBYTE_TOPIC_TEMPLATE = "{workspace} - {connection} - {source} - {destination}"
```
That'll do. To prevent the topic being too long.

aniebietafia commented on zerver/lib/integrations.py:
```diff
@@ -341,6 +341,7 @@ def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
 
 WEBHOOK_INTEGRATIONS: list[WebhookIntegration] = [
     WebhookIntegration("airbrake", ["monitoring"]),
+    WebhookIntegration("airbyte", ["monitoring"], display_name="Airbyte"),
```
Thank you for your comments @laurynmm. I need a little bit of clarity on this phrase `I think our general preference is to have the example fixture data generated directly from the application we're integrating with and not taken from the application's documentation.` Are you saying we need to generate live data directly from Airbyte? Please kindly clarify. Thank you.

aniebietafia replied:
I think I understand what is required here for a real data generated from Airbyte after going through the zulip's [documentation](https://zulip.com/api/incoming-webhooks-walkthrough) again @laurynmm. Here are my thoughts, please correct me if I miss any step.
- I need to use a tool like [Webhook Site]( https://webhook.site/) to capture webhook payload(s) from Airbyte.
- I will then configure Airbyte webhook on Airbyte UI with the URL generated from step 1.
- Run a synchronization job in Airbyte to generate the webhook payload.
- Capture the webhook payload from [webhook.site]( https://webhook.site/) or any other tool.
- Lastly, save this payload as the fixture data.
Please, kindly confirm if I am on the right track.

laurynmm replied:
Yes, that's what I was thinking. You could also use the Zulip [JSON formatter](https://zulip.com/integrations/doc/json) integration to get the payload as well. Just have the URL set up to have the bot send you a DM vs to a channel/topic.

aniebietafia replied:
Thank you. I'm on it now.

aniebietafia replied:
@laurynmm I have been able to generate the data directly from Airbyte but the payload contains a lot more information that may not be needed since the `data` section that I am using. So, I was wondering if I should strip out the unwanted information and use only the section that is needed or leave it as it is for the fixture data. 
The payload looks like this:
```json
{
  "text": "Your connection Google Sheets \u2192 Postgres from Google Sheets to Postgres succeeded\nThis was for null\n\nYou can access its logs here: https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/connections/aa941643-07ea-48a2-9035-024575491720\n\nJob ID: 20441143",
  "blocks": [
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "Sync completed: <https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/connections/aa941643-07ea-48a2-9035-024575491720|Google Sheets \u2192 Postgres>"
      }
    },
    {
      "type": "section",
      "fields": [
        {
          "type": "mrkdwn",
          "text": "*Source:*"
        },
        {
          "type": "mrkdwn",
          "text": "<https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/source/363c0ea3-e989-4051-9f54-d41b794d6621|Google Sheets>"
        },
        {
          "type": "mrkdwn",
          "text": "*Destination:*"
        },
        {
          "type": "mrkdwn",
          "text": "<https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/destination/b3a05072-e3c8-435a-8e6e-4a5c601039c6|Postgres>"
        },
        {
          "type": "mrkdwn",
          "text": "*Duration:*"
        },
        {
          "type": "mrkdwn",
          "text": "1 min 23 sec"
        }
      ]
    },
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "*Sync Summary:*\n1400 record(s) extracted / 1400 record(s) loaded\n281 kB extracted / 281 kB loaded\n"
      }
    }
  ],
  "data": {
    "workspace": {
      "id": "84d2dd6e-82aa-406e-91f3-bf8dbf176e69",
      "name": "Zulip Airbyte Integration",
      "url": "https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69"
    },
    "connection": {
      "id": "aa941643-07ea-48a2-9035-024575491720",
      "name": "Google Sheets → Postgres",
      "url": "https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/connections/aa941643-07ea-48a2-9035-024575491720"
    },
    "source": {
      "id": "363c0ea3-e989-4051-9f54-d41b794d6621",
      "name": "Google Sheets",
      "url": "https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/source/363c0ea3-e989-4051-9f54-d41b794d6621"
    },
    "destination": {
      "id": "b3a05072-e3c8-435a-8e6e-4a5c601039c6",
      "name": "Postgres",
      "url": "https://cloud.airbyte.com/workspaces/84d2dd6e-82aa-406e-91f3-bf8dbf176e69/destination/b3a05072-e3c8-435a-8e6e-4a5c601039c6"
    },
    "jobId": 20441143,
    "startedAt": "2024-10-22T20:27:59Z",
    "finishedAt": "2024-10-22T20:29:22Z",
    "bytesEmitted": 288179,
    "bytesCommitted": 288179,
    "recordsEmitted": 1400,
    "recordsCommitted": 1400,
    "errorMessage": null,
    "durationFormatted": "1 min 23 sec",
    "bytesEmittedFormatted": "281 kB",
    "bytesCommittedFormatted": "281 kB",
    "success": true,
    "durationInSeconds": 83
  }
}
```

laurynmm replied:
No, we want the actual payloads as our fixture data. So it's fine if there is information we're not using in the Zulip message notification.

laurynmm replied:
Also, if you want to use the payload examples below (https://github.com/zulip/zulip/pull/31922#pullrequestreview-2385155158) for the fixture data that would be fine as well.

aniebietafia commented on zerver/webhooks/airbyte/doc.md:
```diff
@@ -0,0 +1,27 @@
+1. {!generate-webhook-url-basic.md!}
+
+1. Go to your project's settings on the Airbyte site. Click on the
+   **Integration** section, and select **Webhook**.
```
You're right, its `Integrations`

pranasziaukas commented on zerver/webhooks/airbyte/fixtures/airbyte_job_payload_failure.json:
```diff
@@ -0,0 +1,37 @@
+{
+    "data": {
+        "workspace": {
+            "id":"b510e39b-e9e2-4833-9a3a-963e51d35fb4",
+            "name":"Workspace1",
+            "url":"https://link/to/ws"
+        },
+        "connection":{
+            "id":"64d901a1-2520-4d91-93c8-9df438668ff0",
+            "name":"Connection",
+            "url":"https://link/to/connection"
+        },
+        "source":{
+            "id":"c0655b08-1511-4e72-b7da-24c5d54de532",
+            "name":"PostgreSQL",
+            "url":"https://link/to/source"
+        },
+        "destination":{
+            "id":"5621c38f-8048-4abb-85ca-b34ff8d9a298",
+            "name":"BigQuery",
+            "url":"https://link/to/destination"
+        },
+        "jobId":9988,
+        "startedAt":"2024-01-01T00:00:00Z",
+        "finishedAt":"2024-01-01T01:00:00Z",
+        "bytesEmitted":1000,
+        "bytesCommitted":1000,
+        "recordsEmitted":89,
+        "recordsCommitted":89,
+        "bytesEmittedFormatted": "1000 B",
+        "bytesCommittedFormatted":"90 B",
+        "success":false,
+        "durationInSeconds":3600,
+        "durationFormatted":"1 hours 0 min",
+        "errorMessage": "Connection timeout while trying to sync data"
+    }
+}
```
Actual payload example:
```suggestion
{
  "text": "Your connection Connection from PostgreSQL to BigQuery just failed...\nThis happened with Checking source connection failed - please review this connection's configuration to prevent future syncs from failing\n\nYou can access its logs here: https://link/to/ws/connections/64d901a1-2520-4d91-93c8-9df438668ff0\n\nJob ID: 20405145",
  "blocks": [
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "Sync failure occurred: <https://link/to/ws/connections/64d901a1-2520-4d91-93c8-9df438668ff0|Connection>"
      }
    },
    {
      "type": "section",
      "fields": [
        {
          "type": "mrkdwn",
          "text": "*Source:*"
        },
        {
          "type": "mrkdwn",
          "text": "<https://link/to/ws/source/c0655b08-1511-4e72-b7da-24c5d54de532|PostgreSQL>"
        },
        {
          "type": "mrkdwn",
          "text": "*Destination:*"
        },
        {
          "type": "mrkdwn",
          "text": "<https://link/to/ws/destination/5621c38f-8048-4abb-85ca-b34ff8d9a298|BigQuery>"
        },
        {
          "type": "mrkdwn",
          "text": "*Duration:*"
        },
        {
          "type": "mrkdwn",
          "text": "14 sec"
        }
      ]
    },
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "*Failure reason:*\n\n```\nChecking source connection failed - please review this connection's configuration to prevent future syncs from failing\n```\n"
      }
    },
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "*Sync Summary:*\n0 record(s) extracted / 0 record(s) loaded\n0 B extracted / 0 B loaded\n"
      }
    }
  ],
  "data": {
    "workspace": {
      "id": "b510e39b-e9e2-4833-9a3a-963e51d35fb4",
      "name": "Workspace1",
      "url": "https://link/to/ws"
    },
    "connection": {
      "id": "64d901a1-2520-4d91-93c8-9df438668ff0",
      "name": "Connection",
      "url": "https://link/to/ws/connections/64d901a1-2520-4d91-93c8-9df438668ff0"
    },
    "source": {
      "id": "c0655b08-1511-4e72-b7da-24c5d54de532",
      "name": "PostgreSQL",
      "url": "https://link/to/ws/source/c0655b08-1511-4e72-b7da-24c5d54de532"
    },
    "destination": {
      "id": "5621c38f-8048-4abb-85ca-b34ff8d9a298",
      "name": "BigQuery",
      "url": "https://link/to/ws/destination/5621c38f-8048-4abb-85ca-b34ff8d9a298"
    },
    "jobId": 20405145,
    "startedAt": "2024-10-22T07:13:48Z",
    "finishedAt": "2024-10-22T07:14:02Z",
    "bytesEmitted": 0,
    "bytesCommitted": 0,
    "recordsEmitted": 0,
    "recordsCommitted": 0,
    "errorMessage": "Checking source connection failed - please review this connection's configuration to prevent future syncs from failing",
    "durationFormatted": "14 sec",
    "bytesEmittedFormatted": "0 B",
    "bytesCommittedFormatted": "0 B",
    "success": false,
    "durationInSeconds": 14
  }
}
```

pranasziaukas commented on zerver/webhooks/airbyte/fixtures/airbyte_job_payload_success.json:
```diff
@@ -0,0 +1,36 @@
+{
+    "data": {
+        "workspace": {
+            "id":"b510e39b-e9e2-4833-9a3a-963e51d35fb4",
+            "name":"Workspace1",
+            "url":"https://link/to/ws"
+        },
+        "connection":{
+            "id":"64d901a1-2520-4d91-93c8-9df438668ff0",
+            "name":"Connection",
+            "url":"https://link/to/connection"
+        },
+        "source":{
+            "id":"c0655b08-1511-4e72-b7da-24c5d54de532",
+            "name":"PostgreSQL",
+            "url":"https://link/to/source"
+        },
+        "destination":{
+            "id":"5621c38f-8048-4abb-85ca-b34ff8d9a298",
+            "name":"BigQuery",
+            "url":"https://link/to/destination"
+        },
+        "jobId":9988,
+        "startedAt":"2024-01-01T00:00:00Z",
+        "finishedAt":"2024-01-01T01:00:00Z",
+        "bytesEmitted":1000,
+        "bytesCommitted":1000,
+        "recordsEmitted":89,
+        "recordsCommitted":89,
+        "bytesEmittedFormatted": "1000 B",
+        "bytesCommittedFormatted":"90 B",
+        "success":true,
+        "durationInSeconds":3600,
+        "durationFormatted":"1 hours 0 min"
+    }
+}
```
Actual payload example:
```suggestion
{
  "text": "Your connection Connection from PostgreSQL to BigQuery succeeded\nThis was for null\n\nYou can access its logs here: https://link/to/ws/connections/64d901a1-2520-4d91-93c8-9df438668ff0\n\nJob ID: 20404614",
  "blocks": [
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "Sync completed: <https://link/to/ws/connections/64d901a1-2520-4d91-93c8-9df438668ff0|Connection>"
      }
    },
    {
      "type": "section",
      "fields": [
        {
          "type": "mrkdwn",
          "text": "*Source:*"
        },
        {
          "type": "mrkdwn",
          "text": "<https://link/to/ws/source/c0655b08-1511-4e72-b7da-24c5d54de532|PostgreSQL>"
        },
        {
          "type": "mrkdwn",
          "text": "*Destination:*"
        },
        {
          "type": "mrkdwn",
          "text": "<https://link/to/ws/destination/5621c38f-8048-4abb-85ca-b34ff8d9a298|BigQuery>"
        },
        {
          "type": "mrkdwn",
          "text": "*Duration:*"
        },
        {
          "type": "mrkdwn",
          "text": "1 min 54 sec"
        }
      ]
    },
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "*Sync Summary:*\n116 record(s) extracted / 116 record(s) loaded\n131 kB extracted / 131 kB loaded\n"
      }
    }
  ],
  "data": {
    "workspace": {
      "id": "b510e39b-e9e2-4833-9a3a-963e51d35fb4",
      "name": "Workspace1",
      "url": "https://link/to/ws"
    },
    "connection": {
      "id": "64d901a1-2520-4d91-93c8-9df438668ff0",
      "name": "Connection",
      "url": "https://link/to/ws/connections/64d901a1-2520-4d91-93c8-9df438668ff0"
    },
    "source": {
      "id": "c0655b08-1511-4e72-b7da-24c5d54de532",
      "name": "PostgreSQL",
      "url": "https://link/to/ws/source/c0655b08-1511-4e72-b7da-24c5d54de532"
    },
    "destination": {
      "id": "5621c38f-8048-4abb-85ca-b34ff8d9a298",
      "name": "BigQuery",
      "url": "https://link/to/ws/destination/5621c38f-8048-4abb-85ca-b34ff8d9a298"
    },
    "jobId": 20404614,
    "startedAt": "2024-10-22T07:05:24Z",
    "finishedAt": "2024-10-22T07:07:18Z",
    "bytesEmitted": 134720,
    "bytesCommitted": 134720,
    "recordsEmitted": 116,
    "recordsCommitted": 116,
    "errorMessage": null,
    "success": true,
    "durationInSeconds": 114,
    "durationFormatted": "1 min 54 sec",
    "bytesEmittedFormatted": "131 kB",
    "bytesCommittedFormatted": "131 kB"
  }
}
```

pranasziaukas reviewed (commented):
In case it's helpful, I'm sharing full Airbyte payloads for both successful and failed sync jobs. I've substituted example dummy names and IDs for privacy reasons, but otherwise these are production-grade payloads generated today.

Also, note that when testing (by pressing the "Test" button next to the webhook input) or saving notification settings, Airbyte sends the following payload:
```json
{
  "text": "Hello World! This is a test from Airbyte to try notification settings for sync successes/failures."
}
```

All in all, for optimal usability and flexibility, I recommend attempting to parse the payload. If parsing fails, fall back to the text field or, as a last resort, output the original JSON. I believe this approach would handle all expected payloads and minimize unexpected behavior in case of breaking changes on Airbyte's side. Something like:
```python
try:
    payload_data = payload["data"]
    # Attempt to extract values using the standard template
    content = AIRBYTE_MESSAGE_TEMPLATE.format(**args)

except Exception:
    content = payload.get("text", str(payload))

return content
```

aniebietafia commented:
@laurynmm please, is type annotation using dict not allowed in zulip's code standards. It seems as that is what is making my solution to fail github actions checks.

aniebietafia commented:
Successfully implemented the latest reviews:
- Generated live data on the Airbyte UI and use it as the fixture data.
- Refactored the code to simplify the message - added an emoji to identify the state of the sync (green for success or red for failure).
- Updated the docs.md file.
- Regenerated the integration static image.
@laurynmm please kindly review whenever you have the time. Thank you.

pranasziaukas commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,95 @@
+
+
+def create_airbyte_message(payload: WildValue) -> tuple[str, str]:
+    data = extract_data_from_payload(payload["data"])
+    content = format_message_from_data(data)
+    topic = create_topic_from_data(data)
```
```suggestion
    if "data" in payload:
        data = extract_data_from_payload(payload["data"])
        content = format_message_from_data(data)
        topic = create_topic_from_data(data)
    else:
        content = payload["text"].tame(check_string)
        topic = "Airbyte"
```

pranasziaukas replied:
This is the simplest implementation I'd be comfortable with personally.

aniebietafia replied:
Hmm. This looks good.

pranasziaukas replied:
> Maybe the fallback topic should be "Airbyte notification" instead of just "Airbyte"?

Agree, @laurynmm.

I actually had `topic = "Airbyte Notification"` originally, but then checked a few random integration implementations and saw that Dropbox had `topic_name = "Dropbox"`. That's the only reason I changed it.

pranasziaukas commented on zerver/webhooks/airbyte/tests.py:
I'd suggest to add `test_airbyte_job_hello_world` with
```python
expected_message = "Hello World! This is a test from Airbyte to try slack notification settings for sync failures."
```

that works with payloads generated when testing/saving Airbyte settings:
```json
{
  "text": "Hello World! This is a test from Airbyte to try slack notification settings for sync failures."
}
```
(actual production payload above)

aniebietafia replied:
Thank you @pranasziaukas for your feedback. Just to be sure that I understand to clearly.
```python
    if "data" in payload:
        data = extract_data_from_payload(payload["data"])
        content = format_message_from_data(data)
        topic = create_topic_from_data(data)
    else:
        content = payload["text"].tame(check_string)
        topic = "Airbyte"
```
provides a fallback option in case of a payload that does not contain `data`.
That is why this test case `test_airbyte_job_hello_world` is necessary, to allow users quickly test their setup without syncing.

laurynmm replied:
Maybe the fallback topic should be "Airbyte notification" instead of just "Airbyte"?

It's definitely a good idea for us to handle the test notification payload, so that way folks know the integration is working as they set it up.

pranasziaukas replied:
@aniebietafia, correct, the suggested code provides a fallback option.

And `test_airbyte_job_hello_world` tests the payloads that Airbyte sends when verifying or saving notification settings. For reference, see the attached screenshot:
<img width="1251" alt="Screenshot 2024-10-24 at 14 06 57" src="https://github.com/user-attachments/assets/dd4e2f53-2d05-43df-ab87-6919f1f498d7">

Without this fallback, users wouldn't have a way to confirm their settings without running full sync jobs, which defeats the purpose of having a "Test" button.

pranasziaukas commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,95 @@
+        "records_committed": payload_data["recordsCommitted"].tame(check_int),
+        "bytes_emitted_formatted": payload_data["bytesEmittedFormatted"].tame(check_string),
+        "bytes_committed_formatted": payload_data["bytesCommittedFormatted"].tame(check_string),
+        "error_message": payload_data.get("errorMessage").tame(check_string)
+        if payload_data.get("errorMessage")
+        else "No error",
```
My understanding is that `payload_data["errorMessage"]` always exists but is `None` when no errors occur. When we use `payload_data.get("errorMessage")`, it returns a falsy value in these cases. We then replace this with “No error,” but this value is never actually used afterward. This results in unnecessarily complicated logic IMO.

pranasziaukas commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,95 @@
+
+
+def extract_data_from_payload(payload_data: WildValue) -> dict[str, str | int | bool]:
+    return {
+        "workspace_name": payload_data["workspace"]["name"].tame(check_string),
+        "connection_name": payload_data["connection"]["name"].tame(check_string),
+        "source_name": payload_data["source"]["name"].tame(check_string),
+        "destination_name": payload_data["destination"]["name"].tame(check_string),
+        "connection_url": payload_data["connection"]["url"].tame(check_string),
+        "source_url": payload_data["source"]["url"].tame(check_string),
+        "destination_url": payload_data["destination"]["url"].tame(check_string),
+        "successful_sync": payload_data["success"].tame(check_bool),
+        "duration_formatted": payload_data["durationFormatted"].tame(check_string),
+        "records_emitted": payload_data["recordsEmitted"].tame(check_int),
+        "records_committed": payload_data["recordsCommitted"].tame(check_int),
+        "bytes_emitted_formatted": payload_data["bytesEmittedFormatted"].tame(check_string),
+        "bytes_committed_formatted": payload_data["bytesCommittedFormatted"].tame(check_string),
+        "error_message": payload_data.get("errorMessage").tame(check_string)
+        if payload_data.get("errorMessage")
+        else "No error",
+    }
```
An optional improvement
```suggestion
    data = {
        "workspace_name": payload_data["workspace"]["name"].tame(check_string),
        "connection_name": payload_data["connection"]["name"].tame(check_string),
        "source_name": payload_data["source"]["name"].tame(check_string),
        "destination_name": payload_data["destination"]["name"].tame(check_string),
        "connection_url": payload_data["connection"]["url"].tame(check_string),
        "source_url": payload_data["source"]["url"].tame(check_string),
        "destination_url": payload_data["destination"]["url"].tame(check_string),
        "successful_sync": payload_data["success"].tame(check_bool),
        "duration_formatted": payload_data["durationFormatted"].tame(check_string),
        "records_emitted": payload_data["recordsEmitted"].tame(check_int),
        "records_committed": payload_data["recordsCommitted"].tame(check_int),
        "bytes_emitted_formatted": payload_data["bytesEmittedFormatted"].tame(check_string),
        "bytes_committed_formatted": payload_data["bytesCommittedFormatted"].tame(check_string),
    }
    if not data["successful_sync"]:
        data["error_message"] = payload_data["errorMessage"].tame(check_string)
    return data
```

pranasziaukas reviewed (changes requested):
This looks pretty good, @aniebietafia!

However, I am confident that we should somehow handle the payloads generated when testing or saving Airbyte settings, as mentioned earlier and detailed below. This approach is necessary to ensure that users can verify their Zulip/Airbyte integration setup without running actual sync jobs.

pranasziaukas reviewed (commented):
LGTM, aside from a small nitpick 👍 🎉

aniebietafia commented on zerver/webhooks/airbyte/view.py:
```diff
@@ -0,0 +1,102 @@
+
+AIRBYTE_TOPIC_TEMPLATE = "{workspace} - {connection}"
+
+AIRBYTE_MESSAGE_TEMPLATE = """
```
That must have sneaked in by mistake 🤦‍♂️. I'll update it immediately

aniebietafia replied:
I'm looking at my code again but that character is not there.
![2024-10-24](https://github.com/user-attachments/assets/ac3f5826-38eb-423f-8faf-91d40ac37ead)

aniebietafia replied:
This is ready for another review as I have implemented the requested changes

pranasziaukas replied:
> I believe this `AIRBYTE_MESSAGE_TEMPLATE = """` does not begin the message with an empty line

I am pretty sure there is an empty newline at the beginning, as I also verified using `assert content.startswith("\n")`. However, since the tests are passing either way, it's likely that the newline is being trimmed somewhere in the process. This is a minor detail that doesn't impact the functionality, so it might not be necessary to worry about.

Other than that, the PR looks good to me, and I have no additional comments 👍

aniebietafia replied:
Thank you @pranasziaukas, I'll make the change then, just in case there are changes in the future.

aniebietafia commented:
Thank you @pranasziaukas, I'll make the change then, just in case there are changes in the future.

aniebietafia commented:
All reviews has been fully implemented. This PR is ready for another review.

laurynmm commented:
@aniebietafia - Thanks for making all of those updates! The notification messages look good and having the test notification messages is great. I noticed a few things to clean up in the documentation and view code, which I pushed here.

One thing I wonder about is the logo shown in the documentation. It'd be nice if they have a logo in their media package without the "Airbyte" text. See how it compares to some of the other integration logos in the main integrations doc page ...

![Screenshot from 2024-10-30 16-25-52](https://github.com/user-attachments/assets/201fdd45-6908-4501-9c86-fe6eef617344)

**Updates**:
- Fixed line breaks in documentation and revised some of the instruction text.
- Updated the view function to generate the message content and topic directly.
- Simplified a few variable names in the message template.

**Updated screenshots**:

<details>
<summary>Updated Airbyte integration page</summary>

![Screenshot from 2024-10-30 16-14-42](https://github.com/user-attachments/assets/f4df87e7-05fb-424d-adef-e0b7f54fcc1d)
</details>

aniebietafia commented:
@laurynmm Thank you for your reviews. I will look up for a logo without the "Airbyte" text in it and make the changes. However, for the 5th point in the documentation, my initial sentence was "Enter the URL generated above in the Webhook URL field. Click Save Changes at the bottom of the page". This is because the `Airbyte settings page` has two Save Changes button, so to avoid users being confused, we point users to the particular button. What do you think?


![frame1](https://github.com/user-attachments/assets/28fa66aa-c9f6-470e-b7af-66dad201af9f)
![frame2](https://github.com/user-attachments/assets/b594cc7c-b17d-450f-a998-cc04b09c84ed)

laurynmm commented:
> However, for the 5th point in the documentation, on the `Airbyte settings page`, there are two `Save Changes` button, so to avoid users being confused as to which `Save Changes` button is being referred to, we point users to the particular button. What do you think?

Ah, yeah, I can see how that might be confusing. In this case, yes let's say which button we're referring to by noting where it is on the page. It might be worth noting that in the commit message for these changes, since we don't normally do that and it's possible someone might change it without looking at the interface in Airbyte. Something like ...

```
Note about the documentation: There are currently two "Save changes"
buttons on the Airbyte "Notifications" settings page, so the
instructions specify which one to use for clarity.
```

Thanks for point that out!

laurynmm commented:
One small nit about the documentation update: You screenshots above show the button in Airbyte as **Save changes**, but the documentation has it as **Save Changes**.

Otherwise, these updates look good to me.

laurynmm commented:
@timabbott - The integration review label is already here, but just wanted to post that this looks good to go for that review.
